src/components/structures/MessagePanel.tsx

```diff
@@ -571,14 +571,38 @@ export default class MessagePanel extends React.Component<IProps, IState> {
         });
     };
 
-    private getNextEventInfo(arr: MatrixEvent[], i: number): { nextEvent: MatrixEvent; nextTile: MatrixEvent } {
+    /**
+     * Find the next event in the list, and the next visible event in the list.
+     *
+     * @param arr        - the list of events to look in
+     * @param shouldShow - which of these events should be shown (a cache of
+     *                     calling this.shouldShowEvent(e) for each e in arr)
+     * @param i          - where in the list we are now
+     *
+     * @returns { nextEvent, nextTile }
+     *
+     * nextEvent is the last event in the supplied array.
+     *
+     * nextTile is the last event in the array that we will show a tile for. It
+     * is used to to determine the 'last successful' flag when rendering the
+     * tile.
+     */
+    private getNextEventInfo(
+        arr: MatrixEvent[],
+        shouldShow: boolean[],
+        i: number,
+    ): { nextEvent: MatrixEvent; nextTile: MatrixEvent } {
+        // WARNING: this method is on a hot path.
+
         const nextEvent = i < arr.length - 1 ? arr[i + 1] : null;
 
-        // The next event with tile is used to to determine the 'last successful' flag
-        // when rendering the tile. The shouldShowEvent function is pretty quick at what
-        // it does, so this should have no significant cost even when a room is used for
-        // not-chat purposes.
-        const nextTile = arr.slice(i + 1).find((e) => this.shouldShowEvent(e));
+        let nextTile = null;
+        for (let n = i + 1; n < arr.length; n++) {
```

Comment:
Why we can't keep the `slice` + `find`  here ?

```ts
const nextTile = arr.slice(i + 1).find((e, n) => shouldShow[n])
```

Reply:
I did this because my first suspicion was that this slice was hurting performance by copying the array. Of course, the JIT might optimise this away.  I did some tests in node...

I switched the implementation in my version (after applying the change you suggested, combining the event and shouldShow info into an object) like this:

```
const nextTile = events.slice(i + 1).find((e) => e.shouldShow)?.event ?? null;
```

I changed my test to run this 1000 times:

```
diff --git a/test/components/structures/MessagePanel-test.tsx b/test/components/structures/MessagePanel-test.tsx
index 7159c18e4d..b180a47189 100644
--- a/test/components/structures/MessagePanel-test.tsx
+++ b/test/components/structures/MessagePanel-test.tsx
@@ -675,7 +675,7 @@ describe("MessagePanel", function () {
         // rendering
 
         const events = [];
-        for (let i = 0; i < 100; i++) {
+        for (let i = 0; i < 1000; i++) {
             events.push(
                 TestUtilsMatrix.mkEvent({
                     event: true,
```

I ran the test 4 times and compared the timings. The version without slice was consistently 10% faster.

If I tidy up the code a bit by writing a utility function, are you happy to keep the no-slice version, or would you like to argue that its simplicity is worth the 10%?

Reply:
Imo, 10% is clearly worth in this case of critical path. So I'm happy to keep the no-slice version :)

Reply:
Maybe worth to add a little comment about the choice to not keep the slice version

Reply:
> Maybe worth to add a little comment about the choice to not keep the slice version

Sorry, I missed this reply but I'll add the comment in my next PR. Thanks!